### PR TITLE
[all] Raise ConnectException only if all hosts in ControllerClient are unreachable

### PR DESCRIPTION
**Pull request by nisargthakkar**
## Raise `ConnectException` only if all hosts in `ControllerClient` are unreachable

To instantiate a URL based ControllerClient, we need to pass in a list of Router or Controller URLs. Since this is a static list that might be configured by operators, it can grow stale over time to include hosts that are no longer running Venice routers or controllers on that port. When querying these endpoints the following cases are possible:
1. A different service is now listening on that port and it is likely to throw a 4XX for route not supported or some 5XX that might trigger some error scenarios in the new service.
2. `java.net.ConnectException` is thrown. This is the most probable option since even though different services might get deployed on that host, the ports are kept mutually exclusive.

In error scenarios, we wish to try our best to provide the user with an actionable error and only return a `ConnectException`
when all hosts throw it. In that case, the `ConnectException` is the most actionable error.

## How was this PR tested?

GH CI

## Does this PR introduce any user-facing changes?

- [X] No. You can skip the rest of this section.
- [ ] Yes. Make sure to explain your proposed changes and call out the behavior change.

### CODE REVIEW

**Review by FelixGV — DISMISSED** [review]

File: internal/venice-test-common/src/integrationtest/java/com/linkedin/venice/controllerapi/TestControllerClient.java [hunk 1]
```diff
@@ -97,8 +97,8 @@ public void testControllerClientWithInvalidUrls() throws IOException {
     String errorResponseStoreName = Utils.getUniqueString("test-error-store");
     String fakeLeaderControllerUrl = "http://fake.leader.controller.url";
 
-    String nonExistentControllerUrl1 = "http://localhost:22";
-    String nonExistentControllerUrl2 = "http://localhost:23";
+    String nonExistentControllerUrl1 = "http://localhost:1100";
+    String nonExistentControllerUrl2 = "http://localhost:1101";
```
Comment (FelixGV) [inline internal/venice-test-common/src/integrationtest/java/com/linkedin/venice/controllerapi/TestControllerClient.java:101]: Must it be a free port? That could make the test flaky if some other test runs at the same time on the host, right...?

Comment (nisargthakkar) [reply internal/venice-test-common/src/integrationtest/java/com/linkedin/venice/controllerapi/TestControllerClient.java:101]: Yes. I assumed that using a free port would automatically make a service listen on it. Tested locally that that was wrong. Updating the PR

Comment (FelixGV) [reply internal/venice-test-common/src/integrationtest/java/com/linkedin/venice/controllerapi/TestControllerClient.java]: Ok... Using `Utils::getFreePort` also has a chance to be flaky BTW, but I guess we could let the test-retry framework take care of that...

On that note... `Utils::getFreePort` doesn't belong in `venice-common`, IMHO. We should probably move it to `TestUtils` in `venice-test-common`...

Comment (nisargthakkar) [reply internal/venice-test-common/src/integrationtest/java/com/linkedin/venice/controllerapi/TestControllerClient.java]: Oh yeah. For sure. I'll move it to `TestUtils`

**Review by FelixGV — APPROVED** [review]
Thanks!

**Pull request merged**